libmonitor: update to version 2018.07.18 and add variant to configure

mwkrentel opened the pull request:
for hpctoolkit.

I have one question.  This PR requires a 1-line change to
cbtf-argonavis and cbtf-krell to specify the old version (no longer
the default or only version).

Should I add that to this PR or create a separate PR?

ping @scheibelp 

ping @jgalarowicz since this affects Open Speedshop.
But the change will be invisible with the two 1-line changes above.

scheibelp commented:
> This PR requires a 1-line change to cbtf-argonavis and cbtf-krell to specify the old version (no longer the default or only version).
> Should I add that to this PR or create a separate PR?

You can add that here.

(also FYI there are flake errors)

mwkrentel commented:
Yes, it spanked me around for the same error in the intel-xed package.
At least I didn't write into column 81.  :-)

I've fixed the flake errors and added the @2013.02.18 version to the
cbtf packages.

scheibelp commented on var/spack/repos/builtin/packages/libmonitor/package.py:
```diff
@@ -26,16 +26,57 @@
+        if '-g' not in flags:
+            flags.append('-g')
+        for flag in flags:
+            if flag[0:2] == '-O':
```
`startswith` would be clearer IMO

scheibelp commented on var/spack/repos/builtin/packages/libmonitor/package.py:
```diff
@@ -26,16 +26,57 @@
+            args.append('--enable-client-signals=%s' % self.signals)
+
+        if '+bgq' in self.spec:
+            args.append('CC=powerpc64-bgq-linux-gcc')
```
I'm slightly worried about how this would interact with Spack's compiler wrappers. Do you have issues when you add an entry to `compilers.yaml`? I'm assuming all dependencies in the DAG should also build with this compiler, and the `CC` setting is to help the `configure` script along. The following should work:

```
if '+bgq' in self.spec:
  args.append('CC=' + spack_cc)
```

(`spack_cc` is provided by Spack to the package, and refers to Spack's compiler wrapper)

mwkrentel commented:
First, I didn't know about startswith, which would explain why I
didn't use it. :-)  You might be right, I'll have to look at this
before I can give you an intelligent answer.

Is this covered somewhere in read the docs?
Or, can you point me to another package that uses this?

As for powerpc64-bgq-linux-gcc, yes, using CC = spack_cc has problems.

The build for hpctoolkit is a bit complicated in places, especially
for blue gene.  We build some parts for the front end and some parts
for the back end.

On blue gene, we build most of the code with the front-end gcc.  But
libmonitor must be built with powerpc64-bgq-linux-gcc (it segfaults
with the front-end gcc).  So, in general, spack_cc will be gcc, but
libmonitor will be different.

Actually, I tried using 'spack compiler find', but I couldn't get it
to accept powerpc64-bgq-linux-gcc as a valid compiler.  Have you tried
this on vulcan?  It didn't work for me, but maybe I was doing
something wrong.

Anyway, this is a one-off that only applies to blue gene and only for
libmonitor (one of the 'quirky' constraints that I warned you about).
And since blue gene is being phased out, I'd rather just stick to the
simple workaround.  This is what we've been doing for years in our old
system (outside of spack).

mwkrentel commented:
Oh, you mean the python string method startswith().  I was picturing
some complicated spack method for dealing with command-line arguments.
Sure, I'll change it.

Sorry, I completely misread what you were saying.

scheibelp commented:
Sorry yes that is what I meant about `startswith`

> On blue gene, we build most of the code with the front-end gcc. But libmonitor must be built with powerpc64-bgq-linux-gcc (it segfaults with the front-end gcc). So, in general, spack_cc will be gcc, but libmonitor will be different.

If I understand correctly, you are saying that the entire libmonitor package builds with the backend compiler? I'm thinking that would be resolved if the libmonitor package in the DAG was assigned the backend compiler and the other specs got the frontend compiler. Right now that's difficult to do since Spack propagates compilers from dependencies up the DAG.

If that sounds like your case, I think it's OK as an interim solution. I'd like to open an issue for you to refer to in a TODO though. Assuming this describes your case, would you mind adding that (when I add the issue)?

> Actually, I tried using 'spack compiler find', but I couldn't get it to accept powerpc64-bgq-linux-gcc as a valid compiler. Have you tried this on vulcan? It didn't work for me, but maybe I was doing something wrong.

Hard to say without more info: did it give an error? Or did it just not add your compiler? Spack may not have searched where that compiler is available - where is it?

mwkrentel commented:
Right.  For hpctoolkit's dependencies on blue gene, we build
everything with front-end gcc, except libmonitor which uses
powerpc64-bgq-linux-gcc (for all of libmonitor).  AFAIK, that's the
only point in 'The Matrix' where we currently do this.

(At least, not counting old Intel Knights-Corner where we use the k1om
cross compiler for 1/2 of the prereqs.  But I'm not trying to support
that here.)

When you say 'refer to in a TODO', did you mean to put a note in the
monitor package.py file?  Or where?

As for spack not finding powerpc64-bgq-linux-gcc, this is what I get
on vulcan.  Am I missing something, or does spack just not understand
the naming convention for cross compilers?

That's an interesting point.  For the IBM compilers, 'xlc' can
sometimes mean the front end, or sometimes the back end.  But GNU
names their cross compilers with the autotools prefix convention,
eg, powerpc64-bgq-linux-gcc or x86_64-k1om-linux-gcc.

On vulcan,

$ type -a powerpc64-bgq-linux-gcc  
powerpc64-bgq-linux-gcc is /usr/local/bin/powerpc64-bgq-linux-gcc

$ spack compiler add powerpc64-bgq-linux-gcc
==> Found no new compilers
==> Compilers are defined in the following files:
    /g/g21/krentel1/.spack/bgq/compilers.yaml

$ spack compiler add /usr/local/bin/powerpc64-bgq-linux-gcc
==> Found no new compilers
==> Compilers are defined in the following files:
    /g/g21/krentel1/.spack/bgq/compilers.yaml

$ spack compiler list
==> Available compilers
-- gcc cnk-ppc64 ------------------------------------------------
gcc@4.4.7

-- gcc rhel6-ppc64 ----------------------------------------------
gcc@4.4.7

-- xl cnk-ppc64 -------------------------------------------------
xl@14.1  xl@12.1

-- xl rhel6-ppc64 -----------------------------------------------
xl@14.1  xl@12.1

-- xl_r cnk-ppc64 -----------------------------------------------
xl_r@14.1  xl_r@12.1

-- xl_r rhel6-ppc64 ---------------------------------------------
xl_r@14.1  xl_r@12.1

scheibelp commented:
> When you say 'refer to in a TODO', did you mean to put a note in the monitor package.py file? Or where?

Exactly. I'll create an issue (or find an existing one) and post it here shortly

> As for spack not finding powerpc64-bgq-linux-gcc, this is what I get on vulcan. Am I missing something, or does spack just not understand the naming convention for cross compilers?

It appears that Spack doesn't understand the compiler naming conventions. It is strict in the types of names it will look for and it currently doesn't account for prefixes like `powerpc64-bgq-linux-gcc`.

scheibelp commented:
So I looked into the precise details on concretization and I think the rules actually would propagate compiler/architecture in a desirable manner. So I think the issue is more with compiler detection than with concretization. I created an issue for that at https://github.com/spack/spack/issues/8860 (and regarding concretization I created an issue at https://github.com/spack/spack/issues/8859).

There may be other details here that I don't fully understand yet: at the moment libmonitor is recorded as a link dependency of `cbtf-argonavis`. How does that package intend to make use of the cross-compiled libmonitor?

mwkrentel commented:
Ok, I added a todo note for the two issues.
Anything else?

scheibelp commented:
> TODO: Spack has trouble with cross-compilers and builds (hpctoolkit) that use both front-end and back-end compilers. See issues #8860 and #8859.

Thanks for adding this! Unfortunately I did a bad job of explaining the issue in 
https://github.com/spack/spack/pull/8834#issuecomment-409747531: this PR is affected by #8860 but not #8859 (I ended up realizing your issue was a bit different while writing #8859). 

Also, it's true that Spack has issues with building a package that uses both front and backend compilers (which we are also working on), but it should actually handle builds where each package is using one compiler (the distinction is a package like boost which builds b2, and if you want to cross compile that you still want to use the frontend compiler to create b2).

That being said I wanted to make sure this was building the way you expect which is why I asked

> libmonitor is recorded as a link dependency of cbtf-argonavis. How does that package intend to make use of the cross-compiled libmonitor?

mwkrentel commented:
I was unsure about which PR, so I put in both.
Did you want me to remove the other one?

I think I checked the box for "allow you to edit my repo."  If it's a
small change and you want to do that, that's ok.

Actually, I did intend to respond to your question about cbtf, I just
have a bunch of things that are more urgent.  But sometime later
today, hopefully.  :-)

scheibelp commented:
> I think I checked the box for "allow you to edit my repo." If it's a small change and you want to do that, that's ok.

Done. I experimented with manually adding a compiler entry and specifying the architecture for the root/dependencies; I found that #8859 does indeed apply here so I did keep a mention of it but added more detail.

mwkrentel commented:
Blue gene is an unusual case of a cross compile.  You can actually
compile some regular, front-end powerpc code and link it into a
program that runs on the back end (but not the other way).  It's more
of a 'cross-link' than a true cross compile where the final link step
must use a back-end wrapper, but some of the pieces can be compiled
with a front-end compiler.

As for cbtf-argonavis, I'm not sure how they combine libmonitor with
the rest of their code, in part because they're using a 5-year old
version.  The default is false, so I left things for their code
exactly as they are now.  If it used to work, then it will continue to
work.

Now, would they benefit from the +bgq option?  Assuming they're
building for the compute nodes, I think so.  But again, it's so old
that I'd have to go back and check.  Probably it wouldn't hurt, but I
don't want to be maintaining their code.  Certainly, they're free to
use it if they want.
